### PR TITLE
add cli command: solana program dump-owned-accounts

### PR DESCRIPTION
**Pull request by mschneider**
#### Problem

In order to run a local test validator with mainnet state to test larger program updates, it's important to be able to import thousands of accounts.

#### Summary of Changes

Added a new subcommand to the cli: program dump-owned-accounts. It creates a directory that contains all program owned accounts in a format that can be imported by solana-test-validator.

Existing subcommand program dump has been renamed to program dump-executable.

### CODE REVIEW

File: cli/src/program.rs [hunk 1]
```diff
@@ -337,8 +343,8 @@ impl ProgramSubCommands for App<'_, '_> {
                         ),
                 )
                 .subcommand(
-                    SubCommand::with_name("dump")
-                        .about("Write the program data to a file")
+                    SubCommand::with_name("dump-executable")
+                        .about("Write the executable program buffer data to a file")
```
Comment (joncinque) [inline cli/src/program.rs:347]: Can you add `.alias("dump")` or keep `"dump"` around with a deprecation notice? Also, can you update the docs at docs/src/cli/deploy-a-program.md to specify using `dump-executable` instead?

File: cli/src/program.rs [hunk 2]
```diff
@@ -356,6 +362,26 @@ impl ProgramSubCommands for App<'_, '_> {
                                 .help("/path/to/program.so"),
                         ),
                 )
+                .subcommand(
+                    SubCommand::with_name("dump-owned-accounts")
+                        .about("Write the program owned accounts to a directory in a format compatible with solana_test_validator")
```
Comment (joncinque) [inline cli/src/program.rs:367]: nit: maybe specify the command name to make things easier? Unless you mean with the crate?
```suggestion
                        .about("Write the program owned accounts to a directory in a format compatible with `solana-test-validator --account-dir`")
```

File: cli/src/program.rs [hunk 3]
```diff
@@ -1478,6 +1515,72 @@ fn process_dump(
...
+            if account.owner == bpf_loader::id()
+                || account.owner == bpf_loader_deprecated::id()
+                || account.owner == bpf_loader_upgradeable::id()
+            {
+                let program_accounts = rpc_client.get_program_accounts(&account_pubkey);
+                match program_accounts {
+                    Ok(program_accounts) => {
+                        for (pubkey, account) in program_accounts {
+                            let output_path =
+                                format!("{}/{}.json", output_directory, pubkey.to_string());
```
Comment (joncinque) [inline cli/src/program.rs:1538]: Can you use a `Path` and use `join` instead here? that way it'll work across platforms

File: cli/src/program.rs [hunk 4]
```diff
@@ -1478,6 +1515,72 @@ fn process_dump(
...
+    if let Some(account_pubkey) = account_pubkey {
+        if let Some(account) = rpc_client
+            .get_account_with_commitment(&account_pubkey, config.commitment)?
+            .value
+        {
+            if account.owner == bpf_loader::id()
+                || account.owner == bpf_loader_deprecated::id()
+                || account.owner == bpf_loader_upgradeable::id()
+            {
+                let program_accounts = rpc_client.get_program_accounts(&account_pubkey);
```
Comment (joncinque) [inline cli/src/program.rs:1533]: How about doing a `fs::create_dir_all` to the `output_directory`? That seems nicer than expecting users to create the directory themselves, unless `File::create` does that automatically

File: cli/src/program.rs [hunk 5]
```diff
@@ -1,3 +1,5 @@
+use solana_cli_output::{CliAccount, CliAccountNewConfig};
```
Comment (joncinque) [inline cli/src/program.rs:1]: nit: I think this will get tripped during CI -- can you fold this into the larger `use` statement below?

**Review by joncinque — COMMENTED** [review]
Thanks for this contribution! It'll be a huge help for local testing

**Comment by mschneider** [comment]
will start working on this sometime february, it's not highest prio on my end, but should find some time eventually

**Comment by joncinque** [comment]
No problem, I'll be on the lookout!

**Pull request closed without merging**